src/main/java/de/hysky/skyblocker/skyblock/dwarven/WishingCompassSolver.java

```diff
@@ -0,0 +1,410 @@
+    }
+
+    public static void onParticle(ParticleS2CPacket packet) {
+        if (!Utils.isInCrystalHollows() || !ParticleTypes.HAPPY_VILLAGER.equals(packet.getParameters().getType())) {
+            return;
+        }
+        //get location of particle
+        Vec3d particlePos = new Vec3d(packet.getX(), packet.getY(), packet.getZ());
+        //update particle used time
+        particleLastUpdate = System.currentTimeMillis();
+        //ignore particle not in the line
+        if (particlePos.distanceTo(particleLastPos) > PARTICLES_MAX_DISTANCE) {
+            return;
+        }
+        particleLastPos = particlePos;
+
+        switch (currentState) {
+            case PROCESSING_FIRST_USE -> {
+                Vec3d particleDirection = particlePos.subtract(startPosOne).normalize();
+                //move direction to fit with particle
+                directionOne = directionOne.add(particleDirection.multiply((double) 1 / PARTICLES_PER_LINE));
+                particleUsedCountOne += 1;
+                //if used enough particle go to next state
+                if (particleUsedCountOne >= PARTICLES_PER_LINE) {
+                    currentState = SolverStates.WAITING_FOR_SECOND;
+                    if (CLIENT.player != null) {
+                        CLIENT.player.sendMessage(Constants.PREFIX.get().append(Text.translatable("skyblocker.config.mining.crystalsWaypoints.wishingCompassSolver.wishingCompassUsedMessage").formatted(Formatting.GREEN)), false);
+                    }
+                }
+            }
+            case PROCESSING_SECOND_USE -> {
+                Vec3d particleDirection = particlePos.subtract(startPosTwo).normalize();
+                //move direction to fit with particle
+                directionTwo = directionTwo.add(particleDirection.multiply((double) 1 / PARTICLES_PER_LINE));
+                particleUsedCountTwo += 1;
+                //if used enough particle go to next state
+                if (particleUsedCountTwo >= PARTICLES_PER_LINE) {
+                    processSolution();
+                }
+            }
+        }
+    }
+
+    private static void processSolution() {
+        if (CLIENT.player == null) {
+            reset();
+            return;
+        }
+        Vec3d targetLocation = solve(startPosOne, startPosTwo, directionOne, directionTwo);
+        if (targetLocation == null) {
+            CLIENT.player.sendMessage(Constants.PREFIX.get().append(Text.translatable("skyblocker.config.mining.crystalsWaypoints.wishingCompassSolver.somethingWentWrongMessage").formatted(Formatting.RED)), false);
+        } else {
+            //send message to player with location and name
+            ZONE playerZone = getZoneOfLocation(startPosOne);
+            MiningLocationLabel.CrystalHollowsLocationsCategory location = getTargetLocation(playerZone);
+            if (!verifyLocation(playerZone, targetLocation)) {
+                location = MiningLocationLabel.CrystalHollowsLocationsCategory.UNKNOWN;
```

Comment:
Should there be an error message sent to the player here?

Reply:
I guess that would make sense to tell the user why its unknown. I shall add this